phoenix-core/src/test/java/org/apache/phoenix/index/IndexToolSingleRowVerifyTest.java

```diff
@@ -0,0 +1,461 @@
+            newCellList.add(newCell);
+            newCellList.add(c);
+            break;
+        case INVALID_CELL_VALUE:
+            if (CellUtil.matchingQualifier(c, EMPTY_COLUMN_BYTES)) {
+                newCell = CellUtil.createCell(CellUtil.cloneRow(c), CellUtil.cloneFamily(c),
+                        Bytes.toBytes(FIRST_VALUE), EnvironmentEdgeManager.currentTimeMillis(),
+                        KeyValue.Type.Put.getCode(),
+                        Bytes.toBytes("zxcv"));
+                newCellList.add(newCell);
+            } else {
+                newCellList.add(c);
+            }
+            break;
+        case INVALID_EMPTY_CELL:
+            if (CellUtil.matchingQualifier(c, EMPTY_COLUMN_BYTES)) {
+                newCell = CellUtil.createCell(CellUtil.cloneRow(c), CellUtil.cloneFamily(c),
+                        CellUtil.cloneQualifier(c), c.getTimestamp(),
+                        KeyValue.Type.Delete.getCode(),
+                        Bytes.toBytes("\\x02"));
+                newCellList.add(newCell);
+            } else {
+                newCellList.add(c);
+            }
+            break;
+        case INVALID_EXTRA_CELL:
+            if (CellUtil.matchingQualifier(c, EMPTY_COLUMN_BYTES)) {
+                newCell = CellUtil.createCell(CellUtil.cloneRow(c), CellUtil.cloneFamily(c),
+                        Bytes.toBytes(FIRST_VALUE), EnvironmentEdgeManager.currentTimeMillis(),
+                        KeyValue.Type.Put.getCode(),
+                        Bytes.toBytes("zxcv"));
+                newCellList.add(newCell);
+            }
+            newCellList.add(c);
+        }
+    }
+
+    private void expireThisRow() {
+        rebuildScanner.indexTableTTL = INDEX_TABLE_EXPIRY_SEC;
+        UnitTestClock expiryClock = new UnitTestClock(5000);
+        EnvironmentEdgeManager.injectEdge(expiryClock);
+    }
+
+    private List<Mutation> getTotalMutationsLessThanExpected(List<Mutation> actualMutations) {
+        Mutation m = actualMutations.get(0);
+        NavigableMap<byte [], List<Cell>> familyCellMap = m.getFamilyCellMap();
+        List<Cell> cellList = familyCellMap.firstEntry().getValue();
+        Cell c = cellList.get(0);
+        Cell newCell = CellUtil.createCell(CellUtil.cloneRow(c), CellUtil.cloneFamily(c),
+                CellUtil.cloneQualifier(c), EnvironmentEdgeManager.currentTimeMillis(),
+                KeyValue.Type.Delete.getCode(),
+                    CellUtil.cloneValue(c));
+        cellList.add(newCell);
+        familyCellMap.put(Bytes.toBytes(0), cellList);
+        m.setFamilyCellMap(familyCellMap);
+        actualMutations.removeAll(actualMutations);
+        actualMutations.add(m);
+        return actualMutations;
+    }
+
+    private void assertVerificationPhaseResult(IndexToolVerificationResult.PhaseResult actualPR,
+            IndexToolVerificationResult.PhaseResult expectedPR) {
+        Assert.assertTrue(actualPR.toString().equalsIgnoreCase(expectedPR.toString()));
```

Comment:
Comparing one PhaseResult to another would be more intuitive and best implemented by overriding equals() instead of comparing output of toString() functions. You wouldn't even need assertVerificationPhaseResult() function then and Assert.Equals() would suffice.